### PR TITLE
webgl: Support texImage2D with a canvas as an argument

### PR DESCRIPTION
**Pull request by emilio**
This involved some refactoring of the 2d context code, which lead to some more test passed there.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7682)

### CODE REVIEW

**Comment by ecoal95** [comment]
Unbitrotted :P

**Comment by ecoal95** [comment]
Wow highfive really hates my rebases, it never removes the label...

cc/ @jdm

File: components/script/dom/htmlcanvaselement.rs [hunk 1]
```diff
@@ -150,6 +150,17 @@ impl LayoutHTMLCanvasElementHelpers for LayoutJS<HTMLCanvasElement> {
 
 
 impl HTMLCanvasElement {
+    pub fn ipc_renderer(&self) -> Option<IpcSender<CanvasMsg>> {
+        if let Some(context) = self.context.get() {
```
Comment (pcwalton) [inline components/script/dom/htmlcanvaselement.rs]: Could use something like

```
self.context.map(|context| {
    match context {
        CanvasContext::Context2d(context) => context.root().r().ipc_renderer(),
        CanvasContext::WebGL(context) => context.root().r().ipc_randerer(),
    }
})
```

Comment (ecoal95) [inline components/script/dom/htmlcanvaselement.rs]: Yes, sure :)

Refactored it also in `LayoutHTMLCanvasElementHelpers`

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 5 of 5 files at r1, 4 of 4 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/canvasrenderingcontext2d.rs, line 962 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7682#-K-3yHxiWrfSt9TF-1f0)** ([raw file](https://github.com/servo/servo/blob/6616f154836a82db4ed403e61b1794f7bc1a4f90/components/script/dom/canvasrenderingcontext2d.rs#L962)):</sup>
Why don't we move this into `fetch_all_data` instead?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7682)

**Comment by ecoal95** [comment]
Review status: 8 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/canvasrenderingcontext2d.rs, line 962 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7682#-K-3yHxiWrfSt9TF-1f0)** ([raw file](https://github.com/servo/servo/blob/6616f154836a82db4ed403e61b1794f7bc1a4f90/components/script/dom/canvasrenderingcontext2d.rs#L962)):</sup>
Because that behaviour is specified for the 2d context API, but not explicitly for WebGL...

Both firefox and chrome seem to return a white bitmap when the canvas has no context, but it's unintuitive something like:

```js
gl.texImage2d(..., canvas_without_context);
canvas_without_context.getContext('webgl'); // <- fails
```

In fact both allow it and the webgl context creation succeeds, I've done the same.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7682)

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/canvasrenderingcontext2d.rs, line 962 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7682#-K-3yHxiWrfSt9TF-1f0)** ([raw file](https://github.com/servo/servo/blob/6616f154836a82db4ed403e61b1794f7bc1a4f90/components/script/dom/canvasrenderingcontext2d.rs#L962)):</sup>
Let's add tests?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7682)

**Comment by ecoal95** [comment]
Review status: 8 of 11 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/canvasrenderingcontext2d.rs, line 962 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7682#-K-3yHxiWrfSt9TF-1f0)** ([raw file](https://github.com/servo/servo/blob/6616f154836a82db4ed403e61b1794f7bc1a4f90/components/script/dom/canvasrenderingcontext2d.rs#L962)):</sup>
Done!

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7682)

**Comment by jdm** [comment]
@bors-servo: r+
Thanks!

---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7682)

**Pull request merged**